sdks/java/core/src/main/java/org/apache/beam/sdk/io/ReadAllViaFileBasedSource.java

```diff
@@ -98,15 +107,53 @@ private ReadFileRangesFn(
     public void process(ProcessContext c) throws IOException {
       ReadableFile file = c.element().getKey();
       OffsetRange range = c.element().getValue();
-      FileBasedSource<T> source =
+      c.output(
           CompressedSource.from(createSource.apply(file.getMetadata().resourceId().toString()))
-              .withCompression(file.getCompression());
+              .withCompression(file.getCompression())
+              .createForSubrangeOfFile(file.getMetadata(), range.getFrom(), range.getTo()));
+    }
+  }
+
+  /**
+   * A {@link Coder} for {@link BoundedSource}s that wraps a {@link SerializableCoder}. We cannot
+   * safely use an unwrapped SerializableCoder because {@link
+   * SerializableCoder#structuralValue(Serializable)} assumes that coded elements support object
+   * equality (https://issues.apache.org/jira/browse/BEAM-3807). By default, Coders compare equality
+   * by serialized bytes, which we want in this case. It is usually safe to depend on coded
+   * representation here because we only compare objects on bundle commit, which compares
+   * serializations of the same object instance.
+   *
+   * <p>BoundedSources are generally not used as PCollection elements, so we do not expose this
+   * coder for wider use.
+   */
+  public static class BoundedSourceCoder<T> extends CustomCoder<BoundedSource<T>> {
+    private final Coder<BoundedSource<T>> coder;
+
+    public BoundedSourceCoder() {
+      coder = (Coder<BoundedSource<T>>) SerializableCoder.of((Class) BoundedSource.class);
+    }
+
+    @Override
+    public void encode(BoundedSource<T> value, OutputStream outStream)
+        throws CoderException, IOException {
+      coder.encode(value, outStream);
+    }
+
+    @Override
+    public BoundedSource<T> decode(InputStream inStream) throws CoderException, IOException {
+      return coder.decode(inStream);
+    }
+  }
+
+  /** Reads elements contained within an input {@link BoundedSource}. */
+  // TODO: Extend to be a Splittable DoFn.
+  public static class ReadFromBoundedSourceFn<T> extends DoFn<BoundedSource<T>, T> {
```

Comment:
Ditto.

Reply:
Wondering the same if I should better let this as it was, because the reuse we do here is pretty low, however is where I noticed the difference between `JavaReadViaImpulse` and `ReadAllViajFileBasedSource`.